File: src/lib/algorithms/kmeans.ts
// Simple k-means for numeric feature vectors
export function kmeans(data: number[][], k: number, iterations = 20) {
  if (!data.length || k <= 0) return { centroids: [], labels: [] }
  const dims = data[0].length
  // init centroids to first k points (simple)
  const centroids = data.slice(0, k).map(d => d.slice())
  const labels = new Array(data.length).fill(0)

  for (let it = 0; it < iterations; it++) {
    // assign
    for (let i = 0; i < data.length; i++) {
      let best = 0
      let bestDist = Infinity
      for (let c = 0; c < k; c++) {
        let dist = 0
        for (let d = 0; d < dims; d++) dist += Math.pow(data[i][d] - centroids[c][d], 2)
        if (dist < bestDist) {
          bestDist = dist
          best = c
        }
      }
      labels[i] = best
    }
    // update
    const sums = Array.from({ length: k }, () => Array(dims).fill(0))
    const counts = Array(k).fill(0)
    for (let i = 0; i < data.length; i++) {
      const c = labels[i]
      counts[c]++
      for (let d = 0; d < dims; d++) sums[c][d] += data[i][d]
    }
    for (let c = 0; c < k; c++) {
      if (counts[c] === 0) continue
      for (let d = 0; d < dims; d++) centroids[c][d] = sums[c][d] / counts[c]
    }
  }

  return { centroids, labels }
}

export default kmeans

File: mobile/lib/algorithms/kmeans.dart
import 'dart:math' as math;

class KMeansResult {
  final List<List<double>> centroids;
  final List<int> labels;
  KMeansResult(this.centroids, this.labels);
}

KMeansResult kmeans(List<List<double>> points, int k, {int maxIter = 50}) {
  if (points.isEmpty || k <= 0) return KMeansResult([], []);
  final dim = points.first.length;
  final rand = math.Random(1234);

  // init centroids randomly
  final centroids = List.generate(k, (_) => List<double>.filled(dim, 0.0));
  for (var i = 0; i < k; i++) {
    centroids[i] = List.from(points[rand.nextInt(points.length)]);
  }

  final labels = List<int>.filled(points.length, 0);

  for (var iter = 0; iter < maxIter; iter++) {
    var changed = false;
    // assign
    for (var i = 0; i < points.length; i++) {
      var best = 0;
      var bestDist = double.infinity;
      for (var c = 0; c < k; c++) {
        var d = 0.0;
        for (var j = 0; j < dim; j++) {
          final diff = points[i][j] - centroids[c][j];
          d += diff * diff;
        }
        if (d < bestDist) {
          bestDist = d;
          best = c;
        }
      }
      if (labels[i] != best) {
        labels[i] = best;
        changed = true;
      }
    }

    // update
    final sums = List.generate(k, (_) => List<double>.filled(dim, 0.0));
    final counts = List<int>.filled(k, 0);
    for (var i = 0; i < points.length; i++) {
      final lbl = labels[i];
      counts[lbl]++;
      for (var j = 0; j < dim; j++) sums[lbl][j] += points[i][j];
    }
    for (var c = 0; c < k; c++) {
      if (counts[c] == 0) continue;
      for (var j = 0; j < dim; j++) centroids[c][j] = sums[c][j] / counts[c];
    }

    if (!changed) break;
  }

  return KMeansResult(centroids, labels);
}
